lib/OpenQA/Cache.pm

```diff
@@ -0,0 +1,182 @@
+package OpenQA::Cache;
+use strict;
+use warnings;
+
+use File::Basename;
+use Fcntl qw(:flock);
+use Mojo::UserAgent;
+use OpenQA::Utils qw(log_error log_info log_debug);
+use OpenQA::Worker::Common;
+use List::MoreUtils;
+use Data::Dumper;
+use JSON;
+
+require Exporter;
+our (@ISA, @EXPORT);
+@ISA    = qw(Exporter);
+@EXPORT = qw(get_asset);
+
+my $cache;
+my $host;
+my $limit   = 50;
+my $remove  = 1;
+my $db_file = "../cache.db";
+
+sub get {
+    my ($asset) = @_;
+    my $index = List::MoreUtils::first_index { $_ eq $asset } @{$cache->{$host}};
+    splice @{$cache->{$host}}, $index, 1 if @{$cache->{$host}} > 0;
+    unshift @{$cache->{$host}}, $asset;
+    expire_asset();
+}
+
+sub set {
+    my ($asset) = @_;
+    unshift @{$cache->{$host}}, $asset;
+    expire_asset();
+}
+
+sub init {
+    my $class;
+    $host = shift;
+    @{$cache->{$host}} = read_db();
+    log_debug(__PACKAGE__ . ": Initialized with $host");
+}
+
+sub update_setup_status {
+    my $id = @_;
+    my $status = {setup => 1};
+    api_call(
+        'post',
+        'jobs/' . $job->{id} . '/status',
+        json     => {status => $status},
+        callback => "no",
+    );
+    log_debug("Update status so job is not considered dead.");
+}
+
+sub download_asset {
+    my ($id, $type, $asset) = @_;
+
+    open(my $log, '>>', "autoinst-log.txt") or die("Cannot open autoinst-log.txt");
+    local $| = 1;
+    print $log "CACHE: Locking $asset";
+
+    print $log "Attemping to download: $host $asset, $type, $id";
+    my $ua = Mojo::UserAgent->new(max_redirects => 2);
+    $ua->max_response_size(0);    #set initial filesize of 20GB
+    my $tx = $ua->build_tx(GET => sprintf '%s/tests/%d/asset/%s/%s', $host, $id, $type, basename($asset));
+
+    $ua->on(
+        start => sub {
+            my ($ua, $tx) = @_;
+            my $progress     = 0;
+            my $last_updated = time;
+            $tx->res->on(
+                progress => sub {
```

Comment:
+1 But actually needs to be a bit more careful on the uploads, as they have a special treatment.